catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/ResultIterable.java

```diff
@@ -212,16 +220,32 @@ private void fetchNextResults() {
             try {
                 SourceResponse response = queryFunction.query(queryRequestCopy);
 
-                if (response.getResults()
-                        .size() == 0) {
+                final List<Result> results = response.getResults();
+                int resultCount = results.size();
+
+                final List<Result> dedupedResults = results.stream()
+                        .filter(this::isDistinctResult)
+                        .collect(Collectors.toList());
+
+                // TODO RAP 19 Sep 17: This test is insufficient, as it does not account
+                // for large blocks of filtered results
+                if (resultCount == 0) {
                     finished = true;
                     return;
                 }
 
-                results = response.getResults()
-                        .iterator();
-                currentIndex += response.getResults()
-                        .size();
+                foundIds.addAll(dedupedResults.stream()
+                        .map(Result::getMetacard)
+                        .filter(Objects::nonNull)
+                        .map(Metacard::getId)
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toSet()));
```

Comment:
❓ Do we want to incur the overhead of the extra set creation instead of just iterating/adding to the `foundResults` set?
Not sure if its worth it, but could be since this could be used very heavily

Reply:
Good suggestion.